### PR TITLE
Introducing LMS Acceleration  in Caliptra 1.1

### PR DESCRIPTION
**Pull request by rusty1968**
This PR integrates Mojtaba's changes by performing light refactoring of the LMS driver to extract the function that performs the hash chain operation and isolating hardware dependencies behind a feature flag. 

It also includes the Winternitz mode emulation and driver tests that enable hw-latest feature.

### CODE REVIEW

**Review by mojtaba-bisheh — APPROVED** [review]

**Review by jlmahowa-amd — DISMISSED** [review]

**Review by JohnTraverAmd — DISMISSED** [review]

File: builder/src/firmware.rs [hunk 1]
```diff
@@ -243,16 +254,31 @@ pub mod driver_tests {
...
+    };
+
     pub const TEST_NEGATIVE_LMS: FwId = FwId {
         bin_name: "test_negative_lms",
         ..BASE_FWID
     };
 
+    pub const TEST_NEGATIVE_LMS_HW_LATEST: FwId = FwId {
+        bin_name: "test_negative_lms_hw_latest",
+        ..BASE_FWID
```
Comment (mhatrevi) [inline builder/src/firmware.rs:279]: Shouldn't this be HW_LATEST_FWID?

File: drivers/src/lms.rs [hunk 2]
```diff
@@ -301,6 +301,68 @@ impl Lms {
...
+        tmp: &mut HashValue<N>,
+    ) -> CaliptraResult<HashValue<N>> {
+        const WNTZ_MODE_SHA256: u8 = 32;
+
+        let iteration_count = ((1u16 << params.w) - 1) as u8;
+
+        if coeff < iteration_count {
+            let mut digest = Array4x8::default();
+            let mut hasher = sha256_driver.digest_init()?;
+            wnt_prefix[22] = coeff;
```
Comment (mhatrevi) [inline drivers/src/lms.rs:320]: There are a lot of hard-coded values in this function. Can we declare constants for these so the program logic is clearer to the reader?

File: builder/src/firmware.rs [hunk 3]
```diff
@@ -152,6 +152,12 @@ pub mod driver_tests {
         features: &["emu"],
     };
 
+    const HW_LATEST_FWID: FwId = FwId {
+        crate_name: "caliptra-drivers-test-bin",
+        bin_name: "",
+        features: &["emu", "hw-latest"],
```
Comment (mhatrevi) [inline builder/src/firmware.rs:158]: Is hw-latest the right feature name for this or should this be something like hw-1.1?

File: drivers/src/lms.rs [hunk 4]
```diff
@@ -301,6 +301,68 @@ impl Lms {
...
+        coeff: u8,
+        params: &LmotsParameter,
+        tmp: &mut HashValue<N>,
+    ) -> CaliptraResult<HashValue<N>> {
+        let iteration_count = ((1u16 << params.w) - 1) as u8;
+
+        for j in coeff..iteration_count {
+            let mut digest = Array4x8::default();
+            let mut hasher = sha256_driver.digest_init()?;
+            wnt_prefix[22] = j;
```
Comment (mhatrevi) [inline drivers/src/lms.rs:353]: See previous comment on hard-coded values.

File: drivers/src/lms.rs [hunk 5]
```diff
@@ -310,7 +372,10 @@ impl Lms {
         y: &[[U32<LittleEndian>; N]; P],
         message_digest: &HashValue<N>,
     ) -> CaliptraResult<HashValue<N>> {
-        let params = get_lmots_parameters(algo_type)?;
+        // wntz_mode: 1 for SHA256 with n=32, and 0 for SHA192 with n=24
```
Comment (mhatrevi) [inline drivers/src/lms.rs:375]: What part of the code is this comment describing?

File: sw-emulator/lib/periph/src/hash_sha256.rs [hunk 6]
```diff
@@ -99,6 +103,18 @@ pub struct HashSha256 {
 
     op_complete_action: Option<ActionHandle>,
 }
+#[derive(Debug)]
+pub struct WntzParams {
+    wntz_mode: bool,
```
Comment (mhatrevi) [inline sw-emulator/lib/periph/src/hash_sha256.rs:108]: Suggest adding descriptions to the Winternitz parameters.

File: sw-emulator/lib/periph/src/hash_sha256.rs [hunk 7]
```diff
@@ -185,6 +183,103 @@ impl HashSha256 {
...
+
+        // If WNTZ_MODE is set and first is set, then enable winternitz
+        if let Ok(params) = self.is_wntz_enabled() {
+            // Initialize winternitz iteration count
+            let iter_count = (1 << params.wntz_w) - 1;
+
+            let mut block = [0u8; 64];
+            block.clone_from(self.block.data());
+            block.to_big_endian();
+            let coeff = block[22] as u16;
```
Comment (mhatrevi) [inline sw-emulator/lib/periph/src/hash_sha256.rs:234]: See previous comment on hard-coded values.

**Review by mhatrevi — CHANGES_REQUESTED** [review]

File: builder/src/firmware.rs (hunk 1 above)
Comment (rusty1968) [reply builder/src/firmware.rs]: Done -: https://github.com/chipsalliance/caliptra-sw/pull/1359/commits/e270e7434bafd7d3fa4ae24e0fe757d271096bb4

File: drivers/src/sha256.rs [hunk 8]
```diff
@@ -235,11 +288,61 @@ impl Sha256 {
         self.digest_op(first)
     }
 
+    /// Calculate digest of the full block
+    ///
+    /// # Arguments
+    ///
+    /// * `block`: Block to calculate the digest
+    /// * `first` - Flag indicating if this is the first block
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:296]: Done : https://github.com/chipsalliance/caliptra-sw/pull/1359/commits/be55039c830816a2d0da1f31efab0671d722bc4a

File: drivers/src/sha256.rs [hunk 9]
```diff
@@ -219,6 +223,55 @@ impl Sha256 {
         Ok(())
     }
 
+    /// Calculate the digest of the last block
+    ///
+    /// # Arguments
+    ///
+    /// * `slice` - Slice of buffer to digest
+    /// * `first` - Flag indicating if this is the first buffer
+    /// * `buf_size` - Total buffer size
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:232]: Done : https://github.com/chipsalliance/caliptra-sw/pull/1359/commits/be55039c830816a2d0da1f31efab0671d722bc4a

File: drivers/src/hmac384.rs [hunk 10]
```diff
@@ -213,8 +224,16 @@ impl Hmac384 {
...
+        cfg_if::cfg_if! {
+            if #[cfg(feature="hw-latest")] {
+                let iv: [u32; 12] = rand_data.0[..12].try_into().unwrap();
+                KvAccess::copy_from_arr(&Array4x12::from(iv), hmac.lfsr_seed())?;
+            }
+            else {
+                let iv: [u32; 5] = rand_data.0[..5].try_into().unwrap();
+                KvAccess::copy_from_arr(&Array4x5::from(iv), hmac.lfsr_seed())?;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/hmac384.rs]: Done

File: drivers/src/lms.rs (hunk 2 above)
Comment (rusty1968) [reply drivers/src/lms.rs]: Done.

File: drivers/src/lms.rs (hunk 4 above)
Comment (rusty1968) [reply drivers/src/lms.rs]: Done

File: drivers/src/lms.rs (hunk 5 above)
Comment (rusty1968) [reply drivers/src/lms.rs]: Removed.

File: drivers/src/sha256.rs [hunk 11]
```diff
@@ -330,6 +473,45 @@ impl<'a> Sha256DigestOp<'a> for Sha256DigestOpHw<'a> {
         Ok(())
     }
 
+    #[cfg(feature = "hw-latest")]
+    /// Update the digest with data
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - Data to used to update the digest
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:487]: Done

File: drivers/src/sha256.rs [hunk 12]
```diff
@@ -235,11 +288,61 @@ impl Sha256 {
...
+            // Submit the first block
+            sha256
+                .ctrl()
+                .write(|w| w.wntz_mode(false).mode(true).init(true).next(false));
+        } else {
+            // Submit next block in existing hashing chain
+            sha256
+                .ctrl()
+                .write(|w| w.wntz_mode(false).mode(true).init(false).next(true));
+        }
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs]: Done

File: drivers/src/sha256.rs [hunk 13]
```diff
@@ -259,6 +362,46 @@ impl Sha256 {
...
         Ok(())
     }
+
+    #[cfg(feature = "hw-latest")]
+    // Perform the digest operation in the hardware
+    //
+    // # Arguments
+    //
+    /// * `first` - Flag indicating if this is the first block
+    fn digest_wntz_op(&mut self, first: bool, w_value: u8, n_mode: bool) -> CaliptraResult<()> {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:370]: Done

File: drivers/src/sha256.rs [hunk 14]
```diff
@@ -353,6 +535,41 @@ impl<'a> Sha256DigestOp<'a> for Sha256DigestOpHw<'a> {
 
         Ok(())
     }
+
+    /// Finalize the digest operations
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:530]: They are not public functions.

File: sw-emulator/lib/periph/src/hash_sha256.rs [hunk 15]
```diff
@@ -34,13 +35,16 @@ register_bitfields! [
             SHA256 = 0b01,
         ],
         ZEROIZE OFFSET(3) NUMBITS(1) [],
-        RSVD OFFSET(4) NUMBITS(28) [],
-    ],
+        WNTZ_MODE OFFSET(4) NUMBITS(1) [],
+        WNTZ_W OFFSET(5)NUMBITS(4) [],
+        WNTZ_N_MODE OFFSET(9) NUMBITS(1) [],
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply sw-emulator/lib/periph/src/hash_sha256.rs:40]: Done

File: sw-emulator/lib/periph/src/hash_sha256.rs (hunk 6 above)
Comment (rusty1968) [reply sw-emulator/lib/periph/src/hash_sha256.rs:110]: Done

File: drivers/src/sha256.rs [hunk 16]
```diff
@@ -259,6 +362,46 @@ impl Sha256 {
...
+    // # Arguments
+    //
+    /// * `first` - Flag indicating if this is the first block
+    fn digest_wntz_op(&mut self, first: bool, w_value: u8, n_mode: bool) -> CaliptraResult<()> {
+        let sha256 = self.sha256.regs_mut();
+
+        // Wait for the hardware to be ready
+        wait::until(|| sha256.status().read().ready());
+
+        if first {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs]: Done

File: sw-emulator/lib/periph/src/hash_sha256.rs (hunk 7 above)
Comment (rusty1968) [reply sw-emulator/lib/periph/src/hash_sha256.rs]: Done

File: drivers/src/sha256.rs [hunk 17]
```diff
@@ -330,6 +473,45 @@ impl<'a> Sha256DigestOp<'a> for Sha256DigestOpHw<'a> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - Data to used to update the digest
+    fn update_wntz(&mut self, data: &[u8], w_value: u8, n_mode: bool) -> CaliptraResult<()> {
+        if self.state == Sha256DigestState::Final {
+            return Err(CaliptraError::DRIVER_SHA256_INVALID_STATE);
+        }
+
+        if self.data_size + data.len() > SHA256_MAX_DATA_SIZE {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs:478]: Done.

File: builder/src/firmware.rs (hunk 3 above)
Comment (rusty1968) [reply builder/src/firmware.rs:158]: This discussion goes beyond the scope of this PR.  The current consensus which  has been approved and merged is to use  the hw-latest switch to point to the RTL release where this feature is present.

File: drivers/src/lms.rs [hunk 18]
```diff
@@ -301,6 +306,66 @@ impl Lms {
         Ok(HashValue::from(digest))
     }
 
+    #[cfg(feature = "hw-latest")]
+    fn hash_chain<const N: usize>(
+        &self,
+        sha256_driver: &mut impl Sha256Alg,
+        wnt_prefix: &mut [u8; 55],
```
Comment (mhatrevi) [inline drivers/src/lms.rs:313]: There are still some hard-coded values. Can we ensure all of these are defined?

File: drivers/src/lms.rs [hunk 19]
```diff
@@ -301,6 +306,66 @@ impl Lms {
...
+        }
+        Ok(*tmp)
+    }
+
+    // This operation is accelerated in hardware by RTL1.1.
+    #[cfg(not(feature = "hw-latest"))]
+    fn hash_chain<const N: usize>(
+        &self,
+        sha256_driver: &mut impl Sha256Alg,
+        wnt_prefix: &mut [u8; 55],
```
Comment (mhatrevi) [inline drivers/src/lms.rs:346]: This values keeps coming up in the code. Let's define an informative constant for this.

File: drivers/src/sha256.rs [hunk 20]
```diff
@@ -235,24 +298,101 @@ impl Sha256 {
...
         wait::until(|| sha256.status().read().ready());
 
-        if first {
-            // Submit the first block
-            sha256.ctrl().write(|w| w.mode(true).init(true).next(false));
-        } else {
-            // Submit next block in existing hashing chain
-            sha256.ctrl().write(|w| w.mode(true).init(false).next(true));
-        }
+        let init = first;
```
Comment (mhatrevi) [inline drivers/src/sha256.rs:334]: Consider simply adding first and !first to init and next.

File: drivers/src/sha256.rs [hunk 21]
```diff
@@ -235,24 +298,101 @@ impl Sha256 {
...
+    //
+    /// * `first` - Flag indicating if this is the first block
+    #[cfg(not(feature = "hw-latest"))]
+    fn digest_op(&mut self, first: bool) -> CaliptraResult<()> {
+        let sha256 = self.sha256.regs_mut();
+
+        // Wait for the hardware to be ready
+        wait::until(|| sha256.status().read().ready());
+
+        let init = first;
```
Comment (mhatrevi) [inline drivers/src/sha256.rs:359]: See previous comment on this.

File: drivers/src/sha256.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (mhatrevi) [reply drivers/src/sha256.rs:486]: This is not fixed.

File: drivers/src/lms.rs (hunk 18 above)
Comment (rusty1968) [reply drivers/src/lms.rs]: done

File: drivers/src/lms.rs (hunk 19 above)
Comment (rusty1968) [reply drivers/src/lms.rs]: done

File: drivers/src/sha256.rs (hunk 21 above)
Comment (rusty1968) [reply drivers/src/sha256.rs]: done

File: drivers/src/sha256.rs [hunk 22]
```diff
@@ -235,24 +298,101 @@ impl Sha256 {
...
+    /// * `first`   - Flag indicating if this is the first block
+    /// * `w_value` - Winternitz W value.
+    /// * `n_mode`  - Winternitz n value(SHA192/SHA256 --> n = 24/32)
+    fn digest_wntz_op(&mut self, first: bool, w_value: u8, n_mode: bool) -> CaliptraResult<()> {
+        let sha256 = self.sha256.regs_mut();
+
+        // Wait for the hardware to be ready
+        wait::until(|| sha256.status().read().ready());
+
+        let init = first;
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/sha256.rs]: done

**Review by mhatrevi — DISMISSED** [review]

File: drivers/Cargo.toml [hunk 23]
```diff
@@ -34,8 +34,14 @@ fpga_realtime = ["caliptra-hw-model/fpga_realtime"]
...
 verilator = ["caliptra-hw-model/verilator"]
 no-cfi = []
+
 hw-latest = ["caliptra-builder/hw-latest", "caliptra-registers/hw-latest"]
 
+
+
+
+
+
```
Comment (jhand2) [inline drivers/Cargo.toml:44]: nit: remove whitespace

File: drivers/src/hmac384.rs [hunk 24]
```diff
@@ -182,6 +172,28 @@ impl Hmac384 {
         Ok(op)
     }
 
+    /// Generate an LFSR seed.
```
Comment (jhand2) [inline drivers/src/hmac384.rs:175]: Can you add a comment about where the seed gets stored after this function is called?

File: drivers/src/sha256.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply drivers/src/sha256.rs:531]: I think this is still missing

File: drivers/test-fw/Cargo.toml [hunk 25]
```diff
@@ -20,9 +20,13 @@ caliptra-cfi-lib = { workspace = true, default-features = false, features = ["cf
 [features]
 emu = ["caliptra-test-harness/emu"]
 fpga_realtime = []
+
 hw-latest = ["caliptra-drivers/hw-latest", "caliptra-registers/hw-latest"]
 
 
+
+
+
```
Comment (jhand2) [inline drivers/test-fw/Cargo.toml:29]: remove extra newlines

File: drivers/test-fw/Cargo.toml [hunk 26]
```diff
@@ -20,9 +20,13 @@ caliptra-cfi-lib = { workspace = true, default-features = false, features = ["cf
 [features]
 emu = ["caliptra-test-harness/emu"]
 fpga_realtime = []
+
```
Comment (jhand2) [inline drivers/test-fw/Cargo.toml:23]: remove

File: drivers/test-fw/Cargo.toml [hunk 27]
```diff
@@ -135,16 +144,34 @@ name = "test_lms_24"
 path = "src/bin/lms_24_tests.rs"
 required-features = ["riscv"]
 
+[[bin]]
+name = "test_lms_24_hw_latest"
+path = "src/bin/lms_24_tests.rs"
+required-features = ["riscv"]
+
```
Comment (jhand2) [inline drivers/test-fw/Cargo.toml:151]: extra newlines here and below

**Review by FerralCoder — APPROVED** [review]

File: drivers/src/hmac384.rs [hunk 28]
```diff
@@ -156,8 +159,16 @@ impl Hmac384 {
 
         // Generate an LFSR seed.
         let rand_data = trng.generate()?;
-        let iv: [u32; 5] = rand_data.0[..5].try_into().unwrap();
-        KvAccess::copy_from_arr(&Array4x5::from(iv), hmac.lfsr_seed())?;
+        cfg_if::cfg_if! {
+            if #[cfg(feature="hw-latest")] {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/hmac384.rs]: Enabling the rtl-1.1 features behind the hw-latest compilation switch is the stop-gap solution that is merged to caliptra-sw.

File: drivers/src/lms.rs [hunk 29]
```diff
@@ -301,6 +301,68 @@ impl Lms {
...
+            hasher.update_wntz(&wnt_prefix[0..23 + N * 4], params.w, n_mode)?;
+            hasher.finalize_wntz(&mut digest, params.w, n_mode)?;
+            *tmp = HashValue::<N>::from(digest);
+        }
+        Ok(*tmp)
+    }
+
+    // This operation is accelerated in hardware by RTL1.1.
+    #[cfg(not(feature = "hw-latest"))]
+    pub fn hash_chain<const N: usize>(
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/lms.rs]: The reason I chose identical  names is there won't be a case where we will not use the accelerated version , as per previous discussions in the firmware meeting.

File: drivers/test-fw/Cargo.toml (hunk 25 above)
Comment (rusty1968) [reply drivers/test-fw/Cargo.toml]: Done

File: drivers/Cargo.toml (hunk 23 above)
Comment (rusty1968) [reply drivers/Cargo.toml]: Done.

File: drivers/src/lms.rs [hunk 30]
```diff
@@ -301,6 +301,68 @@ impl Lms {
...
+                n_mode = true;
+            }
+            hasher.update_wntz(&wnt_prefix[0..23 + N * 4], params.w, n_mode)?;
+            hasher.finalize_wntz(&mut digest, params.w, n_mode)?;
+            *tmp = HashValue::<N>::from(digest);
+        }
+        Ok(*tmp)
+    }
+
+    // This operation is accelerated in hardware by RTL1.1.
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/lms.rs:342]: we do not have information about hw-acceleration version.

File: drivers/src/hmac384.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply drivers/src/hmac384.rs]: Ok, sounds good

File: drivers/src/lms.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply drivers/src/lms.rs]: ack

File: drivers/src/lms.rs (hunk 30 above)
[the remark this answers is not in the document]
Comment (jhand2) [reply drivers/src/lms.rs:342]: I just mean that the comment makes it seem (at least to me) like the implementation below is the hardware accelerated version, but it's not.

[the remark this answers is not in the document]
Comment (rusty1968) [reply drivers/src/lms.rs]: you are right. it was misplaced.

File: drivers/src/hmac384.rs (hunk 24 above)
Comment (rusty1968) [reply drivers/src/hmac384.rs]: Done.

File: drivers/test-fw/Cargo.toml (hunk 27 above)
Comment (rusty1968) [reply drivers/test-fw/Cargo.toml]: removed.

File: drivers/test-fw/Cargo.toml (hunk 26 above)
Comment (rusty1968) [reply drivers/test-fw/Cargo.toml]: removed

**Review by mhatrevi — APPROVED** [review]

**Pull request merged**